libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/QnAMaker/QnAMakerRecognizer.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Licensed under the MIT License.
+// Copyright (c) Microsoft Corporation. All rights reserved.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.AI.QnA;
+using Microsoft.Bot.Expressions;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive.QnA.Recognizers
+{
+    /// <summary>
+    /// IRecognizer implementation which uses QnAMaker KB to identify intents.
+    /// </summary>
+    public class QnAMakerRecognizer : InputRecognizer
+    {
+        [JsonProperty("$kind")]
+        public const string DeclarativeType = "Microsoft.QnAMakerRecognizer";
+
+        public const string QnAMatchIntent = "QnAMatch";
+
+        private Expression knowledgebaseIdExpression;
+        private Expression endpointkeyExpression;
+        private Expression hostnameExpression;
+
+        public QnAMakerRecognizer()
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets the KnowledgeBase Id of your QnA Maker KnowledgeBase.
+        /// </summary>
+        /// <value>
+        /// The knowledgebase Id.
+        /// </value>
+        [JsonProperty("knowledgeBaseId")]
+        public string KnowledgeBaseId
+        {
+            get { return knowledgebaseIdExpression?.ToString(); }
+            set { knowledgebaseIdExpression = value != null ? new ExpressionEngine().Parse(value) : null; }
+        }
+
+        /// <summary>
+        /// Gets or sets the Hostname for your QnA Maker service.
+        /// </summary>
+        /// <value>
+        /// The host name of the QnA Maker knowledgebase.
+        /// </value>
+        [JsonProperty("hostname")]
+        public string HostName
+        {
+            get { return hostnameExpression?.ToString(); }
+            set { hostnameExpression = value != null ? new ExpressionEngine().Parse(value) : null; }
+        }
+
+        /// <summary>
+        /// Gets or sets the Endpoint key for the QnA Maker KB.
+        /// </summary>
+        /// <value>
+        /// The endpoint key for the QnA service.
+        /// </value>
+        [JsonProperty("endpointKey")]
+        public string EndpointKey
+        {
+            get { return endpointkeyExpression?.ToString(); }
+            set { endpointkeyExpression = value != null ? new ExpressionEngine().Parse(value) : null; }
+        }
+
+        /// <summary>
+        /// Gets or sets the number of results you want.
+        /// </summary>
+        /// <value>
+        /// The number of results you want.
+        /// </value>
+        [DefaultValue(3)]
+        [JsonProperty("top")]
+        public int Top { get; set; } = 3;
```

Comment:
Not sure you need both DefaultValue and assign default.  If field is not present in serialization you will get the assigned default.  With what you have here, then from JSON you get the default value and from code the assigned value.  I don't see why you would not want them to be the same.

Reply:
you need both.  If you have the attribute then json.net will not serialize a 3, if you do a new of the object then you want to initialize it.